src/components/TodoItem.tsx

```diff
@@ -0,0 +1,154 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import {
+  FC,
+  ChangeEvent,
+  useState,
+  KeyboardEvent,
+  FormEvent,
+  Dispatch,
+  SetStateAction,
+} from 'react';
+import cn from 'classnames';
+
+import { Todo } from '../types/Todo';
+
+type Props = {
+  todo: Todo;
+  todoToDelete: number[];
+  handleDeleteTodo: (todoId: number) => void;
+  handleUpdateTodo: (
+    todoToUpdate: Todo,
+    isEditing?: boolean,
+    setIsEditing?: Dispatch<SetStateAction<boolean>>,
+  ) => void;
+};
+
+export const TodoItem: FC<Props> = ({
+  todo,
+  todoToDelete,
+  handleDeleteTodo,
+  handleUpdateTodo,
+}) => {
+  const { id, completed, title } = todo;
+
+  const [isEditing, setIsEditing] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+
+  const handleUpdateStatus = (event: ChangeEvent<HTMLInputElement>) => {
+    handleUpdateTodo({
+      ...todo,
+      completed: event.target.checked,
+    });
+  };
+
+  const handleChangeTitle = (event: ChangeEvent<HTMLInputElement>) => {
+    setNewTitle(event.target.value.trimStart());
+  };
+
+  const cancelEditing = () => {
+    setNewTitle(title);
+    setIsEditing(false);
+  };
+
+  const handleCancelEditingOnEsc = (event: KeyboardEvent<HTMLInputElement>) => {
+    if (event.code === 'Escape') {
+      cancelEditing();
+    }
+  };
+
+  const handleSubmit = (
+    event: FormEvent<HTMLFormElement> | React.FocusEvent<HTMLInputElement>,
+  ) => {
+    event.preventDefault();
+
+    if (newTitle.trim() === title) {
+      cancelEditing();
+
+      return;
+    }
+
+    if (!newTitle) {
+      handleDeleteTodo(id);
+
+      return;
+    }
+
+    handleUpdateTodo(
+      {
+        ...todo,
+        title: newTitle.trim(),
+      },
+      isEditing,
+      setIsEditing,
+    );
+  };
+
+  const handleDoubleClick = () => {
+    setIsEditing(true);
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn('todo', {
+        completed: completed,
+      })}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={handleUpdateStatus}
+        />
+      </label>
+
+      {isEditing ? (
+        <form onSubmit={handleSubmit}>
+          <input
+            data-cy="TodoTitleField"
+            type="text"
+            className="todo__title-field"
+            placeholder="Empty todo will be deleted"
+            value={newTitle}
+            onChange={handleChangeTitle}
+            onBlur={handleSubmit}
+            autoFocus
```

Comment:
Using autoFocus in React can lead to a few potential issues:

- Accessibility Concerns: Auto-focusing elements can disrupt users who rely on keyboard navigation or screen readers, making it harder for them to interact with the page.
- Timing Issues: If the component rendering with autoFocus takes time, the focus might not behave as expected, leading to confusion or a poor user experience.

Instead, consider managing focus programmatically using ref and the useEffect hook to ensure better control and timing.